lib/bb/build.py

```diff
@@ -1,395 +1,1057 @@
+    if cleandirs:
+        for cdir in d.expand(cleandirs).split():
+            bb.utils.remove(cdir, True)
+            bb.utils.mkdirhier(cdir)
+
+    if flags and dirs is None:
+        dirs = flags.get('dirs')
+        if dirs:
+            dirs = d.expand(dirs).split()
+
+    if dirs:
+        for adir in dirs:
+            bb.utils.mkdirhier(adir)
+        adir = dirs[-1]
+    else:
+        adir = None
 
-    body = data.getVar(func, d)
+    body = d.getVar(func, False)
     if not body:
+        if body is None:
+            logger.warning("Function %s doesn't exist", func)
         return
 
-    if not dirs:
-        dirs = (data.getVarFlag(func, 'dirs', d) or "").split()
-    for adir in dirs:
-        adir = data.expand(adir, d)
-        mkdirhier(adir)
+    ispython = flags.get('python')
 
-    if len(dirs) > 0:
-        adir = dirs[-1]
+    lockflag = flags.get('lockfiles')
+    if lockflag:
+        lockfiles = [f for f in d.expand(lockflag).split()]
     else:
-        adir = data.getVar('B', d, 1)
-
-    adir = data.expand(adir, d)
+        lockfiles = None
 
-    try:
-        prevdir = os.getcwd()
-    except OSError:
-        prevdir = data.expand('${TOPDIR}', d)
-    if adir and os.access(adir, os.F_OK):
-        os.chdir(adir)
+    tempdir = d.getVar('T')
 
-    if data.getVarFlag(func, "python", d):
-        exec_func_python(func, d)
+    # or func allows items to be executed outside of the normal
+    # task set, such as buildhistory
+    task = d.getVar('BB_RUNTASK') or func
+    if task == func:
+        taskfunc = task
     else:
-        exec_func_shell(func, d)
+        taskfunc = "%s.%s" % (task, func)
+
+    runfmt = d.getVar('BB_RUNFMT') or "run.{func}.{pid}"
+    runfn = runfmt.format(taskfunc=taskfunc, task=task, func=func, pid=os.getpid())
+    runfile = os.path.join(tempdir, runfn)
+    bb.utils.mkdirhier(os.path.dirname(runfile))
+
+    # Setup the courtesy link to the runfn, only for tasks
+    # we create the link 'just' before the run script is created
+    # if we create it after, and if the run script fails, then the
+    # link won't be created as an exception would be fired.
+    if task == func:
+        runlink = os.path.join(tempdir, 'run.{0}'.format(task))
+        if runlink:
+            bb.utils.remove(runlink)
 
-    if os.path.exists(prevdir):
-        os.chdir(prevdir)
+            try:
+                os.symlink(runfn, runlink)
+            except OSError:
+                pass
+
+    with bb.utils.fileslocked(lockfiles):
+        if ispython:
+            exec_func_python(func, d, runfile, cwd=adir)
+        else:
+            exec_func_shell(func, d, runfile, cwd=adir)
 
-def exec_func_python(func, d):
+    try:
+        curcwd = os.getcwd()
+    except:
+        curcwd = None
+
+    if oldcwd and curcwd != oldcwd:
+        try:
+            bb.warn("Task %s changed cwd to %s" % (func, curcwd))
+            os.chdir(oldcwd)
+        except:
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears that you are generically passing an `Exception` object without performing any other operation on it. This may hide error conditions that can otherwise be quickly detected and addressed. We recommend that you catch a more specific exception. If the code must broadly catch all exceptions, consider logging the stack trace using the [logging.exception()](https://docs.python.org/3/library/logging.html#logging.exception) API. For example,

```
try:
    x = 1 / 0
except ZeroDivisionError as e:
    logging.exception(&#39;ZeroDivisionError: %s&#39;, e)

```